File: client/common/protocol/smile_protocol.go
package protocol

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/gcarniglia/tp0-distribuidos/client/common/transport"
)

// Modulo que implementa el Smile Protocol,
// encargado de codificar y decodificar mensajes
type Codec struct{}

func NewCodec() *Codec {
	return &Codec{}
}

func (c *Codec) Encode(msg Message) ([]byte, error) {
	if len(msg.Payload) > MaxPayloadLen {
		return nil, fmt.Errorf("payload too large: %d", len(msg.Payload))
	}
	header := fmt.Sprintf(":)%s %d:(\n", msg.Type, len(msg.Payload))
	buffer := make([]byte, 0, len(header)+len(msg.Payload))
	buffer = append(buffer, []byte(header)...)
	buffer = append(buffer, msg.Payload...)
	return buffer, nil
}

func (c *Codec) DecodeFrom(conn *transport.Conn) (Message, error) {
	headerBytes, err := conn.ReadUntilHeaderTerminator()
	if err != nil {
		return Message{}, err
	}
	header := string(headerBytes)
	if !strings.HasPrefix(header, ":)") || !strings.HasSuffix(header, ":(\n") {
		return Message{}, fmt.Errorf("invalid frame header")
	}

	headerBody := strings.TrimSuffix(strings.TrimPrefix(header, ":)"), ":(\n")
	parts := strings.Split(headerBody, " ")
	if len(parts) != 2 {
		return Message{}, fmt.Errorf("invalid frame header body")
	}

	msgType := parts[0]
	payloadLen, err := strconv.Atoi(parts[1])
	if err != nil || payloadLen < 0 {
		return Message{}, fmt.Errorf("invalid payload length")
	}
	if payloadLen > MaxPayloadLen {
		return Message{}, fmt.Errorf("payload length exceeds max allowed")
	}

	payload, err := conn.ReadExactly(payloadLen)
	if err != nil {
		return Message{}, err
	}

	return Message{Type: MessageType(msgType), Payload: payload}, nil
}

File: server/common/protocol/smile_protocol.py
from server.common.protocol.smile_message import SmileMessage, SmileType
from server.common.protocol.constants import MAX_PAYLOAD_LEN
from server.common.protocol.smile_errors import SmileError, PayloadTooLargeError




'''Modulo que implementa el Smile Protocol, 
encargado de codificar y decodificar mensajes'''
class SmileProtocol:
    def encode(self, message: SmileMessage) -> bytes:
        if len(message.payload) > MAX_PAYLOAD_LEN:
            raise PayloadTooLargeError(len(message.payload))
        header = f":){message.type.value} {len(message.payload)}:(\n".encode("utf-8")
        return header + message.payload

    def decode_from(self, connection) -> SmileMessage:
        header = connection.read_until_header_terminator()
        if not header.startswith(b":)") or not header.endswith(b":(\n"):
            raise SmileError("invalid frame header")

        body = header[2:-3].decode("utf-8")
        parts = body.split(" ")
        if len(parts) != 2:
            raise SmileError("invalid frame header body")

        msg_type = parts[0]
        try:
            payload_len = int(parts[1])
        except ValueError as exc:
            raise SmileError("invalid payload length") from exc

        if payload_len < 0:
            raise SmileError("invalid payload length")

        if payload_len > MAX_PAYLOAD_LEN:
            raise PayloadTooLargeError(payload_len)

        payload = connection.read_exactly(payload_len)
        return SmileMessage(type=SmileType(msg_type), payload=payload)
